src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,546 @@
+                    verrors.add(schema, f'Must be able to communicate with at least one interface on the expansion shelf (node {node}).')
+                    return
+        else:
+            connected_shelf_ips = await self.hardwire_node('', shelf_index, shelf_ip_to_mac)
+            if not connected_shelf_ips:
+                # Failed to connect any IPs => error
+                verrors.add(schema, 'Must be able to communicate with at least one interface on the expansion shelf.')
+                return
+
+    @private
+    async def hardwire_node(self, node, shelf_index, shelf_ip_to_mac):
+        localnode = not node or node == await self.middleware.call('failover.node')
+        shelf_ips = list(shelf_ip_to_mac.keys())
+        # Next see what RDMA-capable links are available on the host
+        # Also setup a map for frequent use below
+        if localnode:
+            links = await self.middleware.call('rdma.get_link_choices')
+        else:
+            links = await self.middleware.call('failover.call_remote', 'rdma.get_link_choices')
+
+        # First check to see if any interfaces that were previously configured
+        # for this shelf are no longer applicable (they might have been moved to
+        # a different port on the JBOF).
+        connected_shelf_ips = set()
+        dirty = False
+        configured_interfaces = await self.middleware.call('rdma.interface.query')
+        configured_interface_names = [interface['ifname'] for interface in configured_interfaces if interface['node'] == node]
+        for interface in configured_interfaces:
+            if node and node != interface['node']:
+                continue
+            host_ip = interface['address']
+            shelf_ip = jbof_static_ip_from_initiator_ip(host_ip)
+            value = decode_static_ip(host_ip)
+            if value and value[0] == shelf_index:
+                # This is supposed to be connected to our shelf.  Check connectivity.
+                if await self.middleware.call('rdma.interface.ping', node, interface['ifname'], shelf_ip, shelf_ip_to_mac[shelf_ip]):
+                    # This config looks good, keep it.
+                    connected_shelf_ips.add(shelf_ip)
+                    if node:
+                        self.logger.info(f'Validated existing link on node {node}: {host_ip} -> {shelf_ip}')
+                    else:
+                        self.logger.info(f'Validated existing link: {host_ip} -> {shelf_ip}')
+                else:
+                    self.logger.info('Removing RDMA interface that cannot connect to JBOF')
+                    await self.middleware.call('rdma.interface.delete', interface['id'])
+                    dirty = True
+        for shelf_ip in shelf_ips:
+            if shelf_ip in connected_shelf_ips:
+                continue
+            # Try each remaining interface
+            if dirty:
+                configured_interfaces = await self.middleware.call('rdma.interface.query')
```

Comment:
Is there a reason to call `rdma.interface.query` for every `shelf_ip` in `shelf_ips` or can this be done once outside the for loop?

Reply:
We need to update `configured_interfaces` when a change is made (`rdma.interface.delete` or `rdma.interface.create`) and this needs to be done before the 
`if ifname not in configured_interface_names:
`

We could refactor to not use the `dirty` variable to control that, but it'll mean we would replace this call site with two.